pystac_client/client.py

```diff
@@ -497,35 +482,28 @@ def get_search_link(self) -> Optional[pystac.Link]:
             None,
         )
 
+    def _get_search_href(self) -> str:
+        search_link = self.get_search_link()
+        href = self._get_href("search", search_link, "search")
+        return href
+
     def _get_collections_href(self, collection_id: Optional[str] = None) -> str:
-        self_href = self.get_self_href()
-        if self_href is None:
-            data_link = self.get_single_link("data")
-            if data_link is None:
-                raise ValueError(
-                    "cannot build a collections href without a self href or a data link"
-                )
-            else:
-                collections_href = data_link.href
+        data_link = self.get_single_link("data")
+        href = self._get_href("data", data_link, "collections")
+        if collection_id is None:
+            return href
         else:
-            collections_href = f"{self_href.rstrip('/')}/collections"
-
-        if not pystac.utils.is_absolute_href(collections_href):
-            collections_href = self._make_absolute_href(collections_href)
+            return f"{href.rstrip('/')}/{collection_id}"
 
-        if collection_id is None:
-            return collections_href
+    def _get_href(self, rel: str, link: Optional[pystac.Link], endpoint: str) -> str:
+        if link and isinstance(link.href, str):
+            href = link.href
+            if not pystac.utils.is_absolute_href(href):
+                href = pystac.utils.make_absolute_href(href, self.self_href)
         else:
-            return f"{collections_href.rstrip('/')}/{collection_id}"
-
-    def _make_absolute_href(self, href: str) -> str:
-        self_link = self.get_single_link("self")
-        if self_link is None:
-            raise ValueError("cannot build an absolute href without a self link")
-        elif not pystac.utils.is_absolute_href(self_link.href):
-            raise ValueError(
-                "cannot build an absolute href from "
-                f"a relative self link: {self_link.href}"
+            warnings.warn(
+                f"No link with {rel=} could be found in this catalog",
+                category=MissingLink,
             )
-        else:
-            return pystac.utils.make_absolute_href(href, self_link.href)
+            href = f"{self.self_href.rstrip('/')}/{endpoint}"
```

Comment:
Not sure what should happen if this endpoint does not exist. Maybe we should ping it or something?

Reply:
My instinct is to not do anything when building the href, as pinging a server every time we make a URL feels not-quite-right. I'd hope that the combination of the warning and the eventual request error would be enough to show the user what went wrong?